cmd/deploy-terraform/terraform.go

```diff
@@ -0,0 +1,183 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/google/shlex"
+	"github.com/opendevstack/ods-pipeline-terraform/internal/command"
+	"github.com/opendevstack/ods-pipeline-terraform/internal/kubernetes"
+	"github.com/opendevstack/ods-pipeline-terraform/internal/output"
+)
+
+const (
+	// exit code returned from terraform plan
+	planSuccessNoChangesExitCode   = 0 // (diff empty)
+	planErrorExitCode              = 1
+	planSuccessWithChangesExitCode = 2 // (diff not empty)
+)
+
+func (d *deployTerraform) terraformCmd(args []string, env map[string]string, dir string, outWriter, errWriter io.Writer) error {
+
+	return command.RunInDir(
+		d.terraformBin, args, env, dir, outWriter, errWriter,
+	)
+}
+
+// terraform plan runs the diff and returns whether the plan is in sync.
+// An error is returned when the plan cannot be started or encounters failures
+func (d *deployTerraform) terraformPlanInSync(args []string, env map[string]string, dir string, outWriter, errWriter io.Writer) (bool, error) {
+	return command.RunWithSpecialFailureCode(
+		d.terraformBin, args, env, dir, outWriter, errWriter, planSuccessWithChangesExitCode,
+	)
+}
+
+func (d *deployTerraform) assembleInitWithK8sBackendArgsEnv() (args []string, env map[string]string, sensitive []string, err error) {
+	args = []string{
+		"init",
+	}
+	// this appears to not be working so we need a backend.tf file.
+	backendConfigs := []string{
+		// "type=kubernetes",
+		// fmt.Sprintf("secret_suffix=%s", d.opts.targetEnvironment),
+		// "in_cluster_config=true",
+	}
+	for _, config := range backendConfigs {
+		args = append(args, fmt.Sprintf("-backend-config=%s", strconv.Quote(config)))
+	}
```

Comment:
Should we remove this part?

Were you able to use those Terraform flags on the CLI and it just doesn't work from Go? Or did those flags also fail to work on the CLI?

Reply:
Yes you are right. Initially I thought this would work, but it did not and I found an issue that said this. So I believe it is not actually supported. But I have not cross checked to see whether I could do this directly from the cli. I will check into this.

Reply:
Here is what I see on the command line:

```
$KUBE_NAMESPACE=test-terraform TF_PLUGIN_CACHE=/tmp/foo TF_VAR_hello=foo terraform init -input=false -backend-config="type=kubernetes" -backend-config="secret_suffix=dev" -no-color

Initializing the backend...

Initializing provider plugins...
- Reusing previous version of hashicorp/tfcoremock from the dependency lock file
- Using previously-installed hashicorp/tfcoremock v0.2.0


Warning: Missing backend configuration

-backend-config was used without a "backend" block in the
configuration.

If you intended to override the default local backend configuration,
no action is required, but you may add an explicit backend block to
your
configuration to clear this warning:

terraform {
  backend "local" {}
}

However, if you intended to override a defined backend, please verify
that
the backend configuration is present and valid.


Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
```

I will for now remove using the `-backend-config` related approach and instead stick with the template approach.